### PR TITLE
Model and Schema update for appLanguage

### LINKED ISSUE #280: Schema and User Model Update for MultiLanguage Support
**Is your feature request related to a problem? Please describe.**
User Model and Schema Modification for app language

**Describe the solution you'd like**
Adding the parameter `appLanguage` in userschema

**Approach to be followed (optional)**
Updating model, schema, and mutation

**Additional context**
This is GSoC related issue.

### PR DESCRIPTION
**Pull request by sumitra19jha**
**What kind of change does this PR introduce?**
Closes #280 

**Did you add tests for your changes?**
No

**If relevant, did you update the documentation?**
Not required

**Summary**
User Model and Schema will be updated so that user language can be stored

**Does this PR introduce a breaking change?**
No

**Other information**

Screenshot 2021-05-26 at 12 56 36 PM

UPDATE LANGUAGE MUTATION

https://user-images.githubusercontent.com/35868598/119628827-14658200-be2b-11eb-9d33-247ad89ccda7.mov

### CODE REVIEW

**Comment by sumitra19jha** [comment]
@palisadoes The PR is ready for the review

**Comment by impiyush83** [comment]
@sumitra19jha  Rather than using bare plain language, I would suggest storing language codes.

https://www.andiamo.co.uk/resources/iso-language-codes/

**Comment by sumitra19jha** [comment]
Nice Suggestion @impiyush83 

Screenshot 2021-05-26 at 2 27 51 PM

Screenshot 2021-05-26 at 2 28 17 PM

**Review by impiyush83 — CHANGES_REQUESTED** [review]

File: resolvers/language_mutation/updateLanguage.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+const User = require('../../models/User');
+
+const updateLanguage = async (parent, args, context) => {
+  // authentication check
+  if (!context.isAuth) throw new Error('User is not authenticated');
```
[the remark this answers is not in the document]
Comment (sumitra19jha) [reply resolvers/language_mutation/updateLanguage.js:5]: context.isAuth also works fine.

[the remark this answers is not in the document]
Comment (impiyush83) [reply resolvers/language_mutation/updateLanguage.js:5]: Functionally it works, but util functions are meant to be resued.

**Review by impiyush83 — APPROVED** [review]

**Comment by palisadoes** [comment]
@impiyush83 @sumitra19jha What's the status on this? Are you in agreement with the approach? Can we review it now?

**Comment by sumitra19jha** [comment]
Yes I checked everything and attached the snapshot. It's working fine

File: resolvers/language_mutation/updateLanguage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sumitra19jha) [reply resolvers/language_mutation/updateLanguage.js:5]: This is a mutation. Not utility. I have used it wherever it is needed

**Review by palisadoes — CHANGES_REQUESTED** [review]
@sumitra19jha This works for now, but I noticed the enum hard codes the number of languages. How will additional languages be added later? Ideally the user should be able to upload language packs for their national languages. How do you propose to do that?

**Comment by sumitra19jha** [comment]
@palisadoes Initially, I thought to use strings for language code, but this can lead to inconsistency on the database level. As talawa app only allows 5 languages currently, so all the mutations made for language should fall within those 5 values only for consistency on the database level. As we increase the number of languages, enums can be added further.

This will help us to make sure that even if we are updating language from let say talawa-admin or in future giving access to API for some additional use, then entry of language code will still be consistent on the database level.

**Comment by sumitra19jha** [comment]
If the number of languages will increase to an extent (lets say 20) then, in that case, we can add a separate enum file for that

**Comment by palisadoes** [comment]
@sumitra19jha What happens if someone in the country of Romania decides to use Talawa? Ideally, they should be able to upload the JSON file to the API database, and the API automatically adds Romanian to the available languages.

The `talawa` app should then get the updated list of languages from the API in the language selection.

What do you think? This way the end users have a lot of flexibility in customizing Talawa to their language needs.

**Comment by sumitra19jha** [comment]
@palisadoes Sure, then I will remove the enum type.

Also, I was thinking that a more efficient way for multilingual support is to `not` involve the backend for the conversion of the app static languages. Currently, the Talawa app has language JSON files. User can choose the language of their choice and the JSON file is responsible for converting app strings into user language preference. 

Visual Studio Code allows us to use google translate API for generating language JSON files. The best benefit of this is that we have to use the API only once in the development process and then the JSON file is attached with the app so that language conversion is not dependent on the backend. We can add as many as 80 languages JSON. My solution was aiming that for all the hardcoded strings we should use the language JSON files.

But Sure, we should provide the admin with the capability of adding their own JSON.

**Comment by palisadoes** [comment]
> @palisadoes Sure, then I will remove the enum type.
> 
> Also, I was thinking that a more efficient way for multilingual support is to `not` involve the backend for the conversion of the app static languages. Currently, the Talawa app has language JSON files. User can choose the language of their choice and the JSON file is responsible for converting app strings into user language preference.
> 
> Visual Studio Code allows us to use google translate API for generating language JSON files. The best benefit of this is that we have to use the API only once in the development process and then the JSON file is attached with the app so that language conversion is not dependent on the backend. We can add as many as 80 languages JSON. My solution was aiming that for all the hardcoded strings we should use the language JSON files.
> 
> But Sure, we should provide the admin with the capability of adding their own JSON.

@sumitra19jha 

My experience with language is that it's difficult to translate single words predictably as there is less context. Words like `run` translate differently in other languages depending on the context. The API will be better for translating chats where the context is easier to determine. Even Instagram has difficulties with this, some of the translations are very bad especially if the grammar of the source material is wrong.

Using the Google API for menus also means additional roaming data charges that less privileged users may not have. I think the Talawa API is better for what we need now. Relying on Google translate for errors is also not advisable unless you also include an error code. 

Getting a translation wrong for menus could be fatal for the acceptance of the app. We have no control over how Google translates from one day to the next. The language specific JSON files help to protect us.

Think of a methodology of using language packs where JSON files can be added dynamically. The app could download the translation from the API when the language is changed. That would be low overhead. The mobile app could download the latest pack each time its started to account for updates.

Then we can consider the Google API for special situations.

What do you think?

**Comment by palisadoes** [comment]
@impiyush83 What are your thoughts on this? It has an impact to some degree on the work you are doing with error handling. I'd prefer to have a standardized method of referring to translations. One method for the API and another for the App is not a good practice.

**Comment by impiyush83** [comment]
@palisadoes  Reading the above comments 

Requirements:

- Talawa App needs to be internationalized.

Scope :

- Only static level translations like menus, fixed text, slogans, etc. ( @palisadoes please confirm the scope ) 

My Thoughts: 

- Most production-ready systems in industry use, localize-like systems that store key:value pairs. But these systems are too costly.  - **Discarded due to very high pricing and network calls **
- Using translation service of a backend for app-specific translations like menus etc is **not a good idea**,  as we are tightly coupling App with Backend. - **Discarded due to tight coupling of Talawa App with Backend **
- Using Google API for the translations also should be avoided as it has network overhead which might slow down the App translations and also have an immense hit on pricing. - **Discarded due to very high pricing and network calls **
-  Talawa App should have an internationalization setup in App itself so that it could cater needs of UI. - **Solution 1**

**Comment by palisadoes** [comment]
@impiyush83 
1. Yes the scope of internationalization would be the menus and related text to start with. Error messages will need to be internationalized too so that DevOps teams in each translated country/region would be able to provide better support.
1. Are you suggesting that all the translation key-value pairs be downloaded when the language changes in the mobile app? 

@sumitra19jha I want to make sure that the same approach is being used in the mobile app, the API and `talawa-admin`.

**Comment by impiyush83** [comment]
@palisadoes  Key-value pairs for all languages should be stored in the mobile app itself with 'English' set as the fallback language.

**Comment by palisadoes** [comment]
> @palisadoes Key-value pairs for all languages should be stored in the mobile app itself with 'English' set as the fallback language.

Thanks, I just wanted to make sure we all had the identical concept in mind to create.

**Comment by sumitra19jha** [comment]
> > @palisadoes Sure, then I will remove the enum type.
> > Also, I was thinking that a more efficient way for multilingual support is to `not` involve the backend for the conversion of the app static languages. Currently, the Talawa app has language JSON files. User can choose the language of their choice and the JSON file is responsible for converting app strings into user language preference.
> > Visual Studio Code allows us to use google translate API for generating language JSON files. The best benefit of this is that we have to use the API only once in the development process and then the JSON file is attached with the app so that language conversion is not dependent on the backend. We can add as many as 80 languages JSON. My solution was aiming that for all the hardcoded strings we should use the language JSON files.
> > But Sure, we should provide the admin with the capability of adding their own JSON.
> 
> @sumitra19jha
> 
> My experience with the language is that it's difficult to translate single words predictably as there is less context. Words like `run` translate differently in other languages depending on the context. Single words like that are often found in menus.
> 
> The API will be better for translating chats where the context is easier to determine. Even Instagram has difficulties with this, some of the translations are very bad especially if the grammar of the source material is wrong.
> 
> Using the Google API for menus also means additional roaming data charges that less privileged users may not have. I think the Talawa API is better for what we need now. Relying on Google translate for errors is also not advisable unless you also include an error code.
> 
> Getting a translation wrong for menus could be fatal for the acceptance of the app. We have no control over how Google translates from one day to the next. The language-specific JSON files help to protect us.
> 
> Think of a methodology of using language packs where JSON files can be added dynamically. The app could download the translation from the API when the language is changed. That would be low overhead. The mobile app could download the latest pack each time it started to account for updates.
> 
> Then we can consider the Google API for special situations.
> 
> What do you think?

@palisadoes I totally agree with you. Language interpretation & conversion are totally dependent on its context. Even with the basic context specified, the translating API is sometimes very bad with the conversion. 

I would also like to bring your attention to the other major problem (also faced by a lot of tech companies) is managing database across multiple languages. Imagine a database with different entries in different languages (for example Some are entering organization name in English other in Spanish, Chinese, Hindi etc). It will be very difficult to keep such a database organized which in turn will affect functionality like search, analytics etc.

We can divide both this problem into three categories and solve each of them independently.

## 1. Strings in App
The constant String used in the app (like app bar name, form field labels, buttons etc), can be handled by adding the JSON language files. VS Code provides a feature in which if `en.json` is specified, then other language JSON files can be created automatically at once. The key features of having this JSON file is that app language conversion wouldn't require any external dependency. This conversion can occur even in the absence of the internet. The key underlying problem in this part is context, some JSON key-value pair contains single words. We have to verify the conversion of those single words to achieve high quality.

## 2.  Communication
An organization registered in Spanish might have the majority of users who knows that language. So communication shouldn't be a problem in the majority of cases. In other cases, we can do the following:

NO CLOUD SERVICE USED
> The user data model contains the language parameter which can let everyone in an organization know the other user's language preference. This can help to facilitate effective communication in case the organization decides not to use the cloud services. Users can know each other language preference and will communicate accordingly.

CLOUD SERVICE USED
> We will provide a translation Key field in organization settings where the user can provide a google translate api key and it will be utilised for language conversion. The problem here is with the security of keeping that key safe. Database inconsistency will be observed as information in multiple languages will be present and also context problems can arise in some cases.

## 3. Dynamic JSON files
If a user wants to add their own language JSON file, then we can provide them with the capability to do that. They can update the language. Here as the user himself or herself will be doing that so there wouldn't be the problem of context. 


As you can see, we cannot completely eradicate both problems. But this can be a good solution as per our use case.

**Comment by palisadoes** [comment]
@sumitra19jha @impiyush83 @yasharth291 

1. We should use a pre-installed JSON file for each language. I leave the method of generation up to you.
2. The `talawa-admin` user should be able to upload a JSON file to override the JSON files installed in the phones.
3. The mobile apps should occasionally check the API for any JSON updates, and download them if necessary.

What languages JSON files will be pre installed in the Dart code, if any?

**Comment by sumitra19jha** [comment]
Language JSON files for all the translation in different languages will be pre-included in Dart. Currently, I have started with only 5 but it will extend up to 80.

**Comment by palisadoes** [comment]
@sumitra19jha 

1. I'll merge this later and create some related issues.
1. No need to do 80 languages. [Focus on the ones in this list, with more than 40 million speakers, total](https://en.m.wikipedia.org/wiki/List_of_languages_by_total_number_of_speakers). That will cover most of the globe.

**Review by palisadoes — APPROVED** [review]

**Comment by palisadoes** [comment]
@Laurell876
@DMills27 
@xsh2047 

Please review this PR

**Review by ranilwallace — APPROVED** [review]
LGTM 👍

**Pull request merged**